synapseclient/core/download/download_async.py

```diff
@@ -326,169 +266,286 @@ class _MultithreadedDownloader:
+        self,
+        url_provider: PresignedUrlProvider,
+        chunk_range_generator: Generator[Tuple[int, int], None, None],
+    ) -> Set[asyncio.Task]:
+        download_tasks = set()
+        session = self._syn._requests_session_storage
+        for chunk_range in chunk_range_generator:
+            start, end = chunk_range
 
-        def session_get():
-            return session.get(
-                presigned_url_provider.get_info().url, headers=range_header
+            download_tasks.add(
+                asyncio.create_task(
+                    self._stream_and_write_chunk_wrapper(
+                        session=session,
+                        url_provider=url_provider,
+                        start=start,
+                        end=end,
+                    )
+                )
             )
 
-        response = None
-        cause = None
-        try:
-            # currently when doing a range request to AWS we retry on anything other than a 206.
-            # this seems a bit excessive (i.e. some 400 statuses would suggest a non-retryable condition)
-            # but for now matching previous behavior.
-            response = with_retry(
-                session_get,
-                expected_status_codes=(HTTPStatus.PARTIAL_CONTENT,),
-                retry_errors=RETRYABLE_CONNECTION_ERRORS,
-                retry_exceptions=RETRYABLE_CONNECTION_EXCEPTIONS,
-            )
-        except Exception as ex:
-            cause = ex
+        return download_tasks
+
+    async def _stream_and_write_chunk_wrapper(
+        self,
+        session: httpx.Client,
+        url_provider: PresignedUrlProvider,
+        start: int,
+        end: int,
+    ) -> Tuple[int, int]:
+        loop = asyncio.get_running_loop()
+        otel_context = context.get_current()
+        return await loop.run_in_executor(
+            self._syn._get_thread_pool_executor(asyncio_event_loop=loop),
+            self._stream_and_write_chunk,
+            session,
+            url_provider,
+            start,
+            end,
+            otel_context,
+        )
 
-        if not response or response.status_code != HTTPStatus.PARTIAL_CONTENT:
-            raise SynapseError(
-                f"Could not download the file: {presigned_url_provider.get_info().file_name},"
-                f" please try again."
-            ) from cause
+    def _check_for_abort(self, start: int, end: int) -> None:
+        """Check if the download has been aborted and raise an exception if so."""
+        with self._thread_lock:
+            if self._aborted:
+                raise SynapseDownloadAbortedException(
+                    f"Download aborted, skipping bytes {start}-{end}"
+                )
 
-        return start, response
+    def _stream_and_write_chunk(
+        self,
+        session: httpx.Client,
+        presigned_url_provider: PresignedUrlProvider,
+        start: int,
+        end: int,
+        otel_context: Union[Context, None],
```

Comment:
do we want to set the default to `otel_context` as `None` here?

Reply:
Since this is a "private" function it doesn't matter much as I am always passing in `Union[Context, None]` in this case. I'll add a doc string to this though